capsules/src/sha256.rs

```diff
@@ -0,0 +1,402 @@
+//! Software implementation of SHA-256.
+//!
+//! Implementation is based on the
+//! Wikipedia description of the algorithm. It performs the hash using
+//! 32-bit native values, translating the input data into the
+//! endianness of the processor and translating the output into
+//! big endian format.
+
+use core::cell::Cell;
+use core::cmp;
+
+use kernel::debug;
+use kernel::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
+use kernel::hil::digest::Client;
+use kernel::hil::digest::{Digest, DigestData, DigestHash, DigestVerify};
+use kernel::utilities::cells::{MapCell, OptionalCell};
+use kernel::utilities::leasable_buffer::LeasableBuffer;
+use kernel::ErrorCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum State {
+    Idle,
+    Data,
+    Hash,
+    Verify,
+}
+
+pub struct Sha256Software<'a> {
+    state: Cell<State>,
+
+    client: OptionalCell<&'a dyn Client<'a, 32>>,
+    input_data: OptionalCell<LeasableBuffer<'static, u8>>,
+    // Used to store the hash or the hash to compare against with verify
+    output_data: Cell<Option<&'static mut [u8; 32]>>,
+
+    hash_values: Cell<[u32; 8]>,
+    round_constants: MapCell<[u32; 64]>,
+
+    deferred_caller: &'a DynamicDeferredCall,
+    handle: OptionalCell<DeferredCallHandle>,
+}
+
+impl<'a> Sha256Software<'a> {
+    pub fn new(call: &'a DynamicDeferredCall) -> Sha256Software<'a> {
+        Sha256Software {
+            state: Cell::new(State::Idle),
+            client: OptionalCell::empty(),
+            input_data: OptionalCell::empty(),
+            output_data: Cell::new(None),
+
+            hash_values: Cell::new([0; 8]),
+            round_constants: MapCell::new([0; 64]),
+
+            deferred_caller: call,
+            handle: OptionalCell::empty(),
+        }
+    }
+
+    pub fn initialize_callback_handle(&self, handle: DeferredCallHandle) {
+        self.handle.replace(handle);
+    }
+
+    pub fn busy(&self) -> bool {
+        match self.state.get() {
+            State::Idle => false,
+            _ => true,
+        }
+    }
+
+    pub fn initialize(&self) -> Result<(), ErrorCode> {
+        if !self.busy() {
+            self.hash_values.set([
+                0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab,
+                0x5be0cd19,
+            ]);
+            self.round_constants.map(|k| {
```

Comment:
Why not set this in `new()`? Do we expect to call `initialize()` multiple times? That's worth documenting then

Reply:
Because it needs to be invoked in `clear_data`. Initialization of the hash values is a per-hash operation.  I could factor out the round constants to `new` or make them const data. The latter probably makes more sense/?

Reply:
Changed it to const data.